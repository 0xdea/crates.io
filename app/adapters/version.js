import ApplicationAdapter from './application';

export default class VersionAdapter extends ApplicationAdapter {
  urlForUpdateRecord(id, modelName, snapshot) {
    let crateName = snapshot.record.crate.id;
    let num = snapshot.record.num;
    return `/${this.namespace}/crates/${crateName}/${num}`;
  }

  urlForQueryRecord(query) {
    let { name, num } = query ?? {};
    let baseUrl = this.buildURL('crate', name);
    let url = `${baseUrl}/${num}`;
    // The following used to remove them from URL's query string.
    delete query.name;
    delete query.num;
    return url;
  }
}
