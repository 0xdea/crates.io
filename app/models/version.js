import DS from 'ember-data';

export default DS.Model.extend({
    num: DS.attr('string'),
    dl_path: DS.attr('string'),
    crate: DS.belongsTo('crate'),
    created_at: DS.attr('date'),
    updated_at: DS.attr('date'),
    downloads: DS.attr('number'),
});
