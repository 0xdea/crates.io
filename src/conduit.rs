extern crate semver;

use std::collections::HashMap;
use std::io::net::ip::IpAddr;

#[deriving(PartialEq, Show, Clone)]
pub enum Scheme {
    Http,
    Https
}

#[deriving(PartialEq, Show, Clone)]
pub enum Host<'a> {
    HostName(&'a str),
    HostIp(IpAddr)
}

#[deriving(PartialEq, Show, Clone)]
pub enum Method<'a> {
    Get,
    Post,
    Put,
    Delete,
    Head,
    Connect,
    Options,
    Trace,

    // RFC-5789
    Patch,
    Purge,

    // WebDAV, Subversion, UPNP
    Other(&'a str)
}

pub trait Request {
    /// The version of HTTP being used
    fn http_version(&self) -> semver::Version;

    /// The version of the conduit spec being used
    fn conduit_version(&self) -> semver::Version;

    /// The request method, such as GET, POST, PUT, DELETE or PATCH
    fn method<'a>(&'a self) -> Method<'a>;

    /// The scheme part of the request URL
    fn scheme(&self) -> Scheme;

    /// The host part of the requested URL
    fn host<'a>(&'a self) -> Host<'a>;

    /// The initial part of the request URL's path that corresponds
    /// to a virtual root. This allows an application to have a
    /// virtual location that consumes part of the path.
    fn virtual_root<'a>(&'a self) -> Option<&'a str>;

    /// The remainder of the path.
    fn path<'a>(&'a self) -> &'a str;

    /// The portion of the request URL that follows the "?"
    fn query_string<'a>(&'a self) -> Option<&'a str>;

    /// The remote IP address of the client or the last proxy that
    /// sent the request.
    fn remote_ip(&self) -> IpAddr;

    /// The byte-size of the body, if any
    fn content_length(&self) -> Option<uint>;

    /// The request's headers, as conduit::Headers.
    fn headers<'a>(&'a self) -> &'a Headers;

    /// A Reader for the body of the request
    fn body<'a>(&'a mut self) -> &'a mut Reader;
}

pub type HeaderEntries<'a> = Box<Iterator<(&'a str, Vec<&'a str>)>>;

pub trait Headers {
    /// Find the value of a given header. Multi-line headers are represented
    /// as an array.
    fn find<'a>(&'a self, key: &str) -> Option<Vec<&'a str>>;

    /// Returns true if a particular header exists
    fn has(&self, key: &str) -> bool;

    /// Iterate over all of the available headers.
    fn iter<'a>(&'a self) -> HeaderEntries<'a>;
}

pub struct Response {
    /// The status code as a tuple of the return code and status string
    pub status: (uint, &'static str),

    /// A Map of the headers
    pub headers: HashMap<String, Vec<String>>,

    /// A Writer for body of the response
    pub body: Box<Reader + Send>
}

/// A Handler takes a request and returns a response or an error.
/// By default, a bare function implements `Handler`.
pub trait Handler<E> {
    fn call(&self, request: &mut Request) -> Result<Response, E>;
}

impl<E> Handler<E> for fn(&mut Request) -> Result<Response, E> {
    fn call(&self, request: &mut Request) -> Result<Response, E> {
        (*self)(request)
    }
}
