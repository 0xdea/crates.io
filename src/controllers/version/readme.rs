use crate::app::AppState;
use crate::controllers::version::CrateVersionPath;
use crate::util::{RequestUtils, redirect};
use axum::response::{IntoResponse, Response};
use axum_extra::json;
use http::request::Parts;

/// Get the readme of a crate version.
#[utoipa::path(
    get,
    path = "/api/v1/crates/{name}/{version}/readme",
    params(CrateVersionPath),
    tag = "versions",
    responses((status = 200, description = "Successful Response")),
)]
pub async fn get_version_readme(app: AppState, path: CrateVersionPath, req: Parts) -> Response {
    let redirect_url = app.storage.readme_location(&path.name, &path.version);
    if req.wants_json() {
        json!({ "url": redirect_url }).into_response()
    } else {
        redirect(redirect_url)
    }
}
