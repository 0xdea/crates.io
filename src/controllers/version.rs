pub mod downloads;
pub mod metadata;
pub mod yank;

use axum::extract::{FromRequestParts, Path};
use diesel_async::AsyncPgConnection;
use serde::de::Error;
use serde::{Deserialize, Deserializer};
use utoipa::IntoParams;

use crate::models::{Crate, Version};
use crate::util::errors::{crate_not_found, AppResult};

#[derive(Deserialize, FromRequestParts, IntoParams)]
#[into_params(parameter_in = Path)]
#[from_request(via(Path))]
pub struct CrateVersionPath {
    /// Name of the crate
    pub name: String,
    /// Version number
    #[param(example = "1.0.0")]
    #[serde(deserialize_with = "deserialize_version")]
    pub version: String,
}

impl CrateVersionPath {
    pub async fn load_version(&self, conn: &mut AsyncPgConnection) -> AppResult<Version> {
        Ok(self.load_version_and_crate(conn).await?.0)
    }

    pub async fn load_version_and_crate(
        &self,
        conn: &mut AsyncPgConnection,
    ) -> AppResult<(Version, Crate)> {
        version_and_crate(conn, &self.name, &self.version).await
    }
}

async fn version_and_crate(
    conn: &mut AsyncPgConnection,
    crate_name: &str,
    semver: &str,
) -> AppResult<(Version, Crate)> {
    use diesel::prelude::*;
    use diesel_async::RunQueryDsl;

    let krate: Crate = Crate::by_name(crate_name)
        .first(conn)
        .await
        .optional()?
        .ok_or_else(|| crate_not_found(crate_name))?;

    let version = krate.find_version(conn, semver).await?;

    Ok((version, krate))
}

fn deserialize_version<'de, D: Deserializer<'de>>(deserializer: D) -> Result<String, D::Error> {
    let s = String::deserialize(deserializer)?;
    let _ = semver::Version::parse(&s).map_err(Error::custom)?;
    Ok(s)
}
