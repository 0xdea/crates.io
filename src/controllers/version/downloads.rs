//! Functionality for downloading crates and maintaining download counts
//!
//! Crate level functionality is located in `krate::downloads`.

use super::CrateVersionPath;
use crate::app::AppState;
use crate::models::VersionDownload;
use crate::schema::*;
use crate::util::errors::{version_not_found, AppResult};
use crate::util::{redirect, RequestUtils};
use crate::views::EncodableVersionDownload;
use axum::response::{IntoResponse, Response};
use axum_extra::json;
use axum_extra::response::ErasedJson;
use chrono::{Duration, NaiveDate, Utc};
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use http::request::Parts;

/// Download a crate version.
///
/// This returns a URL to the location where the crate is stored.
#[utoipa::path(
    get,
    path = "/api/v1/crates/{name}/{version}/download",
    tag = "versions",
    responses((status = 200, description = "Successful Response")),
)]
pub async fn download_version(
    app: AppState,
    path: CrateVersionPath,
    req: Parts,
) -> AppResult<Response> {
    let wants_json = req.wants_json();
    let redirect_url = app.storage.crate_location(&path.name, &path.version);
    if wants_json {
        Ok(json!({ "url": redirect_url }).into_response())
    } else {
        Ok(redirect(redirect_url))
    }
}

/// Get the download counts for a crate version.
///
/// This includes the per-day downloads for the last 90 days.
#[utoipa::path(
    get,
    path = "/api/v1/crates/{name}/{version}/downloads",
    tag = "versions",
    responses((status = 200, description = "Successful Response")),
)]
pub async fn get_version_downloads(
    app: AppState,
    path: CrateVersionPath,
    req: Parts,
) -> AppResult<ErasedJson> {
    if semver::Version::parse(&path.version).is_err() {
        return Err(version_not_found(&path.name, &path.version));
    }

    let mut conn = app.db_read().await?;
    let (version, _) = path.load_version_and_crate(&mut conn).await?;

    let cutoff_end_date = req
        .query()
        .get("before_date")
        .and_then(|d| NaiveDate::parse_from_str(d, "%F").ok())
        .unwrap_or_else(|| Utc::now().date_naive());
    let cutoff_start_date = cutoff_end_date - Duration::days(89);

    let downloads = VersionDownload::belonging_to(&version)
        .filter(version_downloads::date.between(cutoff_start_date, cutoff_end_date))
        .order(version_downloads::date)
        .load(&mut conn)
        .await?
        .into_iter()
        .map(VersionDownload::into)
        .collect::<Vec<EncodableVersionDownload>>();

    Ok(json!({ "version_downloads": downloads }))
}
