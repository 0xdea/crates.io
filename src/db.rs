use std::fmt::Show;
use std::mem;
use std::sync::Arc;

use pg;
use pg::{PostgresConnection, PostgresStatement, PostgresResult};
use pg::types::ToSql;
use r2d2::{mod, LoggingErrorHandler};
use r2d2_postgres::PostgresPoolManager;
use conduit::{Request, Response};
use conduit_middleware::Middleware;

use app::{App, RequestApp};
use {user, package, version};
use util::{CargoResult, LazyCell};

pub type Pool = r2d2::Pool<pg::PostgresConnection,
                           pg::error::PostgresConnectError,
                           PostgresPoolManager,
                           LoggingErrorHandler>;
type PooledConnnection<'a> =
        r2d2::PooledConnection<'a,
                               pg::PostgresConnection,
                               pg::error::PostgresConnectError,
                               PostgresPoolManager,
                               LoggingErrorHandler>;

pub fn pool(url: &str, config: r2d2::Config) -> Pool {
    let mgr = PostgresPoolManager::new(url, pg::NoSsl);
    r2d2::Pool::new(config, mgr, LoggingErrorHandler).unwrap()
}

pub fn setup(conn: &PostgresConnection) {
    user::setup(conn);
    package::setup(conn);
    version::setup(conn);
}

pub struct TransactionMiddleware;

pub struct Transaction {
    // fields are destructed top-to-bottom so ensure we destroy them in the
    // right order.
    tx: LazyCell<pg::PostgresTransaction<'static>>,
    slot: LazyCell<PooledConnnection<'static>>,

    // Keep a handle to the app which keeps a handle to the database to ensure
    // that this `'static` is indeed at least a little more accurate (in that
    // it's alive for the period of time this `Transaction` is alive.
    _app_handle: Arc<App>,
}

impl Middleware for TransactionMiddleware {
    fn before(&self, req: &mut Request) -> Result<(), Box<Show + 'static>> {
        let app = req.app().clone();
        req.mut_extensions().insert(Transaction {
            _app_handle: app,
            slot: LazyCell::new(),
            tx: LazyCell::new(),
        });
        Ok(())
    }

    fn after(&self, req: &mut Request, res: Result<Response, Box<Show + 'static>>)
             -> Result<Response, Box<Show + 'static>> {
        match res {
            Ok(ref res) if res.status.val0() == 200 => {
                let tx = req.extensions().find::<Transaction>()
                            .expect("Transaction not present in request");
                match tx.tx.borrow() {
                    Some(tx) => {
                        if req.app().config.env != ::Test {
                            tx.set_commit();
                        }
                    }
                    None => {}
                }
            }
            _ => {}
        }
        return res;
    }
}

pub trait RequestTransaction<'a> {
    /// Return the lazily initialized postgres connection for this request.
    ///
    /// The connection will live for the lifetime of the request.
    fn db_conn(self) -> CargoResult<&'a PostgresConnection>;

    /// Return the lazily initialized postgres transaction for this request.
    ///
    /// The transaction will live for the duration of the request, and it will
    /// only be set to commit() if a successful response code of 200 is seen.
    fn tx(self) -> CargoResult<&'a pg::PostgresTransaction<'a>>;
}

impl<'a> RequestTransaction<'a> for &'a Request + 'a {
    fn db_conn(self) -> CargoResult<&'a PostgresConnection> {
        let tx = self.extensions().find::<Transaction>()
                     .expect("Transaction not present in request");
        // Here we want to tie the lifetime of a single connection the lifetime
        // of this request. Currently the lifetime of a connection is tied to
        // the lifetime of the pool from which it came from, which is the
        // mismatch.
        //
        // The unsafety here is frobbing lifetimes to ensure that they work out.
        // Additionally, any extension in a Request needs to live for the static
        // lifetime (yay!).
        //
        // To solve these problems, the private `Transaction` extension stores a
        // reference both to the pool (to keep it alive) as well as a connection
        // transmuted to the 'static lifetime. This allows us to allocate a
        // connection up front and then repeatedly hand it out.
        unsafe {
            if !tx.slot.filled() {
                let conn: PooledConnnection = try!(self.app().database.get());
                tx.slot.fill(mem::transmute::<_, PooledConnnection<'static>>(conn));
            }
        }
        Ok(&**tx.slot.borrow().unwrap())
    }

    fn tx(self) -> CargoResult<&'a pg::PostgresTransaction<'a>> {
        let tx = self.extensions().find::<Transaction>()
                     .expect("Transaction not present in request");
        // Similar to above, the transaction for this request is actually tied
        // to the connection in the request itself, not 'static. We transmute it
        // to static as its paired with the inner connection to achieve the
        // desired effect.
        unsafe {
            if !tx.tx.filled() {
                let conn = try!(self.db_conn());
                let t = try!(conn.transaction());
                let t = mem::transmute::<_, pg::PostgresTransaction<'static>>(t);
                tx.tx.fill(t);
            }
        }
        let tx = tx.tx.borrow();
        let tx: &'a pg::PostgresTransaction<'static> = tx.unwrap();
        Ok(tx)
    }
}

pub trait Connection {
    fn prepare<'a>(&'a self, query: &str) -> PostgresResult<PostgresStatement<'a>>;
    fn execute(&self, query: &str, params: &[&ToSql]) -> PostgresResult<uint>;
}

impl Connection for pg::PostgresConnection {
    fn prepare<'a>(&'a self, query: &str) -> PostgresResult<PostgresStatement<'a>> {
        self.prepare(query)
    }
    fn execute(&self, query: &str, params: &[&ToSql]) -> PostgresResult<uint> {
        self.execute(query, params)
    }
}
//
// impl Connection for pg::pool::PooledPostgresConnection {
//     fn prepare<'a>(&'a self, query: &str) -> PostgresResult<PostgresStatement<'a>> {
//         self.prepare(query)
//     }
//     fn execute(&self, query: &str, params: &[&ToSql]) -> PostgresResult<uint> {
//         self.execute(query, params)
//     }
// }

impl<'a> Connection for pg::PostgresTransaction<'a> {
    fn prepare<'a>(&'a self, query: &str) -> PostgresResult<PostgresStatement<'a>> {
        self.prepare(query)
    }
    fn execute(&self, query: &str, params: &[&ToSql]) -> PostgresResult<uint> {
        self.execute(query, params)
    }
}
