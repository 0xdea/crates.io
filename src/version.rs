use std::collections::{HashSet, HashMap};
use time::Timespec;

use conduit::{Request, Response};
use conduit_router::RequestParams;
use pg::PostgresRow;
use pg::types::ToSql;
use semver;
use url;

use app::{App, RequestApp};
use db::{Connection, RequestTransaction};
use package::Package;
use util::{RequestUtils, CargoResult, Require, internal, CargoError};
use util::errors::NotFound;

#[deriving(Clone)]
pub struct Version {
    pub id: i32,
    pub package_id: i32,
    pub num: String,
    pub updated_at: Timespec,
    pub created_at: Timespec,
}

#[deriving(Encodable, Decodable)]
pub struct EncodableVersion {
    pub id: i32,
    pub pkg: String,
    pub num: String,
    pub url: String,
    pub updated_at: String,
    pub created_at: String,
}

impl Version {
    pub fn from_row(row: &PostgresRow) -> Version {
        Version {
            id: row.get("id"),
            package_id: row.get("package_id"),
            num: row.get("num"),
            updated_at: row.get("updated_at"),
            created_at: row.get("created_at"),
        }
    }

    pub fn find(conn: &Connection, version_id: i32)
                -> CargoResult<Version> {
        let stmt = try!(conn.prepare("SELECT * FROM versions \
                                      WHERE id = $1"));
        let mut rows = try!(stmt.query(&[&version_id]));
        match rows.next().map(|r| Version::from_row(&r)) {
            Some(version) => Ok(version),
            None => Err(NotFound.box_error()),
        }
    }

    pub fn find_by_num(conn: &Connection, package_id: i32, num: &str)
                       -> CargoResult<Option<Version>> {
        let stmt = try!(conn.prepare("SELECT * FROM versions \
                                      WHERE package_id = $1 AND num = $2"));
        let mut rows = try!(stmt.query(&[&package_id, &num as &ToSql]));
        Ok(rows.next().map(|r| Version::from_row(&r)))
    }

    pub fn insert(conn: &Connection, package_id: i32,
                  num: &str) -> CargoResult<Version> {
        let stmt = try!(conn.prepare("INSERT INTO versions \
                                      (package_id, num, updated_at, \
                                       created_at, downloads) \
                                      VALUES ($1, $2, $3, $4, 0) \
                                      RETURNING *"));
        let now = ::now();
        let mut rows = try!(stmt.query(&[&package_id, &num as &ToSql, &now, &now]));
        Ok(Version::from_row(&try!(rows.next().require(|| {
            internal("no version returned")
        }))))
    }

    pub fn valid(version: &str) -> bool {
        semver::Version::parse(version).is_ok()
    }

    pub fn encodable(self, app: &App, pkg: &Package) -> EncodableVersion {
        let Version { id, package_id, num, updated_at, created_at } = self;
        assert_eq!(pkg.id, package_id);
        EncodableVersion {
            url: format!("https://{}{}",
                         app.bucket.host(), pkg.path(num.as_slice())),
            num: num,
            id: id,
            pkg: pkg.name.clone(),
            updated_at: ::encode_time(updated_at),
            created_at: ::encode_time(created_at),
        }
    }
}

pub fn index(req: &mut Request) -> CargoResult<Response> {
    let conn = try!(req.tx());

    // Extract all ids requested.
    let query = url::form_urlencoded::parse_str(req.query_string().unwrap_or(""));
    let ids = query.iter().filter_map(|&(ref a, ref b)| {
        if a.as_slice() == "ids[]" {
            from_str(b.as_slice())
        } else {
            None
        }
    }).collect::<Vec<i32>>();

    // Load all versions
    //
    // TODO: can rust-postgres do this for us?
    let mut versions = Vec::new();
    let mut set = HashSet::new();
    let query = format!("'{{{:#}}}'::int[]", ids.as_slice());
    let stmt = try!(conn.prepare(format!("SELECT * FROM versions \
                                          WHERE id = ANY({})",
                                         query).as_slice()));
    for row in try!(stmt.query(&[])) {
        let v = Version::from_row(&row);
        set.insert(v.package_id);
        versions.push(v);
    }

    // Load all packages
    let ids = set.into_iter().collect::<Vec<i32>>();
    let query = format!("'{{{:#}}}'::int[]", ids.as_slice());
    let stmt = try!(conn.prepare(format!("SELECT * FROM packages \
                                          WHERE id = ANY({})",
                                         query).as_slice()));
    let mut map = HashMap::new();
    for row in try!(stmt.query(&[])) {
        let p = Package::from_row(&row);
        map.insert(p.id, p);
    }

    // And respond!
    let versions = versions.into_iter().map(|v| {
        let id = v.package_id;
        v.encodable(&**req.app(), map.find(&id).unwrap())
    }).collect();

    #[deriving(Encodable)]
    struct R { versions: Vec<EncodableVersion> }
    Ok(req.json(&R { versions: versions }))
}

pub fn show(req: &mut Request) -> CargoResult<Response> {
    let id = &req.params()["version_id"];
    let id = from_str(id.as_slice()).unwrap_or(0);
    let conn = try!(req.tx());
    let version = try!(Version::find(&*conn, id));
    let pkg = try!(Package::find(&*conn, version.package_id));

    #[deriving(Encodable)]
    struct R { version: EncodableVersion }
    Ok(req.json(&R { version: version.encodable(&**req.app(), &pkg) }))
}
