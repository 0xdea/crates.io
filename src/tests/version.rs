use std::collections::HashMap;
use serialize::json;

use conduit::{mod, Handler, Request};
use semver;

use cargo_registry::db::RequestTransaction;
use cargo_registry::version::{EncodableVersion, Version};

#[deriving(Decodable)]
struct VersionList { versions: Vec<EncodableVersion> }
#[deriving(Decodable)]
struct VersionResponse { version: EncodableVersion }

fn sv(s: &str) -> semver::Version {
    semver::Version::parse(s).unwrap()
}

#[test]
fn index() {
    let (_b, app, middle) = ::app();
    let mut req = ::req(app, conduit::Get, "/versions");
    let mut response = ok_resp!(middle.call(&mut req));
    let json: VersionList = ::json(&mut response);
    assert_eq!(json.versions.len(), 0);

    let (v1, v2) = {
        ::mock_user(&mut req, ::user("foo"));
        let c = ::mock_crate(&mut req, ::krate("foo"));
        let req = &mut req as &mut Request;
        let tx = req.tx().unwrap();
        let m = HashMap::new();
        let v1 = Version::insert(tx, c.id, &sv("2.0.0"), &m, []).unwrap();
        let v2 = Version::insert(tx, c.id, &sv("2.0.1"), &m, []).unwrap();
        (v1, v2)
    };
    req.with_query(format!("ids[]={}&ids[]={}", v1.id, v2.id));
    let mut response = ok_resp!(middle.call(&mut req));
    let json: VersionList = ::json(&mut response);
    assert_eq!(json.versions.len(), 2);
}

#[test]
fn show() {
    let (_b, app, middle) = ::app();
    let mut req = ::req(app, conduit::Get, "/versions");
    let v = {
        ::mock_user(&mut req, ::user("foo"));
        let krate = ::mock_crate(&mut req, ::krate("foo"));
        let req = &mut req as &mut Request;
        let tx = req.tx().unwrap();
        Version::insert(tx, krate.id, &sv("2.0.0"), &HashMap::new(), []).unwrap()
    };
    req.with_path(format!("/versions/{}", v.id).as_slice());
    let mut response = ok_resp!(middle.call(&mut req));
    let json: VersionResponse = ::json(&mut response);
    assert_eq!(json.version.id, v.id);
}

#[test]
fn authors() {
    let (_b, app, middle) = ::app();
    let mut req = ::req(app, conduit::Get, "/crates/foo/1.0.0/authors");
    ::mock_user(&mut req, ::user("foo"));
    ::mock_crate(&mut req, ::krate("foo"));
    let mut response = ok_resp!(middle.call(&mut req));
    let s = response.body.read_to_string().unwrap();
    let json = json::from_str(s.as_slice()).unwrap();
    let json = json.as_object().unwrap();
    assert!(json.contains_key(&"users".to_string()));
}
