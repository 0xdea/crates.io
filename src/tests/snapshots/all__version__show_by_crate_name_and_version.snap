---
source: src/tests/version.rs
assertion_line: 82
expression: json
---
version:
  audit_actions: []
  crate: foo_vers_show
  crate_size: 1234
  created_at: "[datetime]"
  dl_path: /api/v1/crates/foo_vers_show/2.0.0/download
  downloads: 0
  features: {}
  id: "[id]"
  license: ~
  links:
    authors: /api/v1/crates/foo_vers_show/2.0.0/authors
    dependencies: /api/v1/crates/foo_vers_show/2.0.0/dependencies
    version_downloads: /api/v1/crates/foo_vers_show/2.0.0/downloads
  num: 2.0.0
  published_by:
    avatar: ~
    id: "[id]"
    login: foo
    name: ~
    url: "https://github.com/foo"
  readme_path: /api/v1/crates/foo_vers_show/2.0.0/readme
  updated_at: "[datetime]"
  yanked: false

